Improve code (and fix some memory leaks)

ema-pe opened the pull request:
Hi, I read your [blog post](https://schneems.com/2017/06/14/meditations-on-writing-a-queue/), and I found some errors. I'm not a C expert, so you shouldn't totally trust me!

Every commit has an explanation, so I don't need to rewrite they here, but I have some advices:

1. I don't know Ruby, but in C you should use four whitespace instead of two.
2. Remember that C has not a garbage collector, so you need to deallocate manually the allocated memory! Remember also that `malloc` might return NULL.

All examples contained memory leaks, because you didn't deallocate the memory of queue structure. I've fixed them, but a memory leak still remains in `hello_struct.c` because you need ad-hoc function to deallocate structures.

I repeat, I'm not a C expert, I'm a beginner as you, so take these advices with care.

Commit:
Add error checking on malloc

From malloc's man page: "The  malloc() and calloc() functions return a
pointer to the allocated memory, which is suitably aligned for any
built-in type. On error, these functions return NULL."

So I've changed the functions tiny_queque_create and tiny_queque_push.
The latter returns -1 if malloc didn't allocate memory, while the first
retuns NULL.

I haven't changed the examples, but they should be updated.

Commit:
Add error checking for 'mutex' and 'wakeup'

From pthread_cond_init's man page: "In cases where default condition
variable attributes are appropriate, the macro PTHREAD_COND_INITIALIZER
can be used to initialize condition variables that are statically
allocated. The effect shall be equivalent to dynamic initialization by a
call to pthread_cond_init() with parameter attr specified as NULL,
except that no error checks are performed.

If successful, the pthread_cond_destroy() and pthread_cond_init()
functions shall return zero; otherwise, an error number shall be
returned to indicate the error."

The same logic has pthread_mutex_init, so insteadi of using a macro I've used
directly the function, with error checking. The last two statements are
executed only if the mutex or conditional variable is not set up
correcty, so I deallocate the memory.

Commit:
Decrease critical section of tiny_queue_push

It is known that the software is faster if the critical sections are
very small. So I've moved the lock of 'mutex' some lines after, because
malloc and the set up of the new node doesn't require to lock the entire
queue.

Commit:
Fix memory leak by adding tiny_queue_destroy

You need to deallocate the queue when you have finished to use it; if
you don't it you have a memory leak. So I've added a new function,
'tiny_queue_destroy', that deallocate all elements of the queue and the
queue structure itself.

Now the examples 'hello.c' and 'hello_thread.c' works correcly, while
'hello_struct.c' needs other coding, because you can't use free() to
deallocate recursively a structure (so the memory leak on this code
still remains).

Commit:
Replace AND to OR operator

schneems commented:
Thanks! I'm in the weeds right now but I'll try to get to this as soon as possible. I appreciate the feedback and the PR!

Commit:
The AND operator is necessary

Commit:
Fix memory leak in 'hello_struct.c'

It is a good habit to handle structures with a pair of functions that
works similary to the constructor and destroyer of OOP.

ema-pe commented:
I've corrected the memory leak located in `hello_struct.c`, but if the queue contains elements the leak still remains. Fell free to ask any question, especially about `city_destroy` function.

schneems commented:
Thanks!
